recipes/paho-mqtt-cpp/all/patches/1.0.1/0004-fix-win32-static-config-cmake.patch

```diff
@@ -0,0 +1,17 @@
+diff --git a/cmake/FindPahoMqttC.cmake b/cmake/FindPahoMqttC.cmake
+index 53f15a3..7e2d756 100644
+--- a/cmake/FindPahoMqttC.cmake
++++ b/cmake/FindPahoMqttC.cmake
+@@ -6,8 +6,10 @@ else()
+     set(_PAHO_MQTT_C_LIB_NAME paho-mqtt3a)
+ endif()
+ # add suffix when using static Paho MQTT C library variant
+-if(PAHO_BUILD_STATIC)
+-    set(_PAHO_MQTT_C_LIB_NAME ${_PAHO_MQTT_C_LIB_NAME}-static)
++if(WIN32)
++       if(PAHO_BUILD_STATIC)
++       set(_PAHO_MQTT_C_LIB_NAME ${_PAHO_MQTT_C_LIB_NAME}-static)
++       endif()
```

Comment:
This does not match

https://github.com/conan-io/conan-center-index/blob/85f47df4cfc365f14c000051260e77dcf7545972/recipes/paho-mqtt-c/all/conanfile.py#L123-L124

can you explain why you added this patch?

Reply:
1.3.8 paho-mqtt-c build is removing -static from the library name for non WIN32 builds. 
[Example 1.3.8](https://github.com/eclipse/paho.mqtt.c/blob/317fb008e1541838d1c29076d2bc5c3e4b6c4f53/src/CMakeLists.txt#L154)

This patch is to be able to build older versions of paho-mqtt-cpp with version 1.3.8 of paho-mqtt-c and to make sure the library name is correct for non WIN32 platforms.
[paho-mqtt-cpp FindPahoMqtt](https://github.com/eclipse/paho.mqtt.cpp/blob/33921c8b68b351828650c36816e7ecf936764379/cmake/FindPahoMqttC.cmake#L10)

The target concat with -static should still work.

Reply:
#3903 leave's this in a grey zone....

I could not image there's an expectation from consumer that older `-cpp`s will work with newer `-c`s

Instead of a patch I would suggestion having a conditional requirements... and adding a `validate()` method to make sure the overrides are correct

Does not sound reasonable to you?

Reply:
The build for paho-mqtt-cpp was broken for non x86_64 platforms. The patch will also fix building on other architectures. I am a bit new to conan so I don't know the correct path forward to fix non x86_64 broken builds with older -cpp consumers. 

> Instead of a patch I would suggestion having a conditional requirements... and adding a validate() method to make sure the overrides are correct

Do you have an example where this has been done?

I also don't know how long the non x86_64 builds have been broken, as I was unable to build anything in the 1.3 series of paho-mqtt-c before 1.3.8 with paht-mqtt-cpp consuming it with the following exception with a specific recipe; paho-mqtt-c/1.3.5@#acc9376a7232c5a50ac898a9131c6c82, which caused conan to want to run with --update and I am wanting to pin versions.

Reply:
Would patching this be allowed under #3903 for the following:

>They allow libraries to build in other configurations.

Reply:
> The patch will also fix building on other architectures

I am not very familiar with these libraries, how is the `-static` suffix impacting non x86_64?

> Do you have an example where this has been done?

Absolutely we do!

https://github.com/conan-io/conan-center-index/blob/4116c939117a47bf4c4285aeec2a857e6b9b7165/recipes/spdlog/all/conanfile.py#L52-L58

and the validate method can have something like this

https://github.com/conan-io/conan-center-index/blob/4116c939117a47bf4c4285aeec2a857e6b9b7165/recipes/spdlog/all/conanfile.py#L90-L91

***

Looking at the recents closed PRs https://github.com/conan-io/conan-center-index/pull/2739 I think this is a complicated issue...

the `paho-mqtt-c` exported CMake targets are different from the one's which `poha-mqtt-cpp` is trying to consume 😖 

Let's try with matching the correct versions together, but it might not be so simple

Reply:
> I am not very familiar with these libraries, how is the -static suffix impacting non x86_64?

Version 1.2.0 fixed the static suffix for non WIN32 builds shown here [1.2.0 fix -static suffxx](https://github.com/eclipse/paho.mqtt.cpp/commit/6e87712bf4f356d398ce9945fb270b92bf877db0#diff-682732893dd0f12e8753a2f33367bd61cc747f0599b577819adad9ad2474b6af). The patches I added were to fix this for older versions. Non x86 builds do not have a downloadable binary from the conan index repo and have to be built from source. The static suffix for older versions was expected to be there but wasn't on non WIN32 builds so building from source was failing. See #3760

Creating a validate() method and fixing up requirements might be fairly straight forward, but a large majority of static build configuration not having a pre-built paho-mqtt-c library available on conan center would trigger ConanInvalidConfiguration.

> the paho-mqtt-c exported CMake targets are different from the one's which poha-mqtt-cpp is trying to consume

Does the namespace handle this?  [https://github.com/eclipse/paho.mqtt.c/blame/e047e25d34d53b4b265649144a3cac3b01eee76c/src/CMakeLists.txt#L308](https://github.com/eclipse/paho.mqtt.c/blame/e047e25d34d53b4b265649144a3cac3b01eee76c/src/CMakeLists.txt#L308)

Reply:
You should delete the file you are trying to patch. And change the call to `find_package` to use the official one you pointed to

> Why?

Conan works replicated the standard files from build systems, for CMake using theses generators 
https://github.com/conan-io/conan-center-index/blob/58739e044c9e34d6baae695d6d9019fecec19d4a/recipes/paho-mqtt-cpp/all/conanfile.py#L14

Will create the correct `Find<Package>.cmake` files

https://github.com/conan-io/conan-center-index/blob/58739e044c9e34d6baae695d6d9019fecec19d4a/recipes/paho-mqtt-c/all/conanfile.py#L93

➡️ But I do not like this either

***

## The Problem

`paho-mqtt-cpp` does **not** use official/correct cmake generated files from `paho-mqtt-c`

> Does the namespace handle this?

No, the `-cpp` project is completely skipping the `-c` files generation... the owner explained his reasoning .... sadly best practices have changed

***

## Solution

- Fix `paho-mqtt-c`

https://github.com/eclipse/paho.mqtt.c/blob/v1.3.8/src/CMakeLists.txt#L154

does not match the Code in this repository as I pointed our here https://github.com/conan-io/conan-center-index/pull/4096#discussion_r555440378

- Fix `paho-mqtt-cpp`

self.version < 1.2.0
- `-c` MUST be less then 1.3.8
else
- >= 1.3.8

Reply:
> paho-mqtt-cpp does not use official/correct cmake generated files from paho-mqtt-c
> No, the -cpp project is completely skipping the -c files generation... the owner explained his reasoning .... sadly best practices have changed

Being a conan build newb, I am trying to understand how to know if the -cpp uses the correct -c cmake files. Is the following correct?
The -c package_info is being called by the -cpp build and the cmake Find file is being placed in the -cpp build directory. The -cpp build then uses those generated cmake find files to locate the eclipse-paho-mqtt-c library files.

Is it the generators in the -cpp recipe doing this?  - generators = "cmake", "cmake_find_package"
I can get it to build with generators = "cmake", and this doesn't generate the Findeclipse-paho-mqtt-c.cmake file. Is there more going on?

My changes on the -cpp conanfile.py currently looks like this:

```
diff --git a/recipes/paho-mqtt-cpp/all/conanfile.py b/recipes/paho-mqtt-cpp/all/conanfile.py
index 7fe996cb..63a7c2a9 100644
--- a/recipes/paho-mqtt-cpp/all/conanfile.py
+++ b/recipes/paho-mqtt-cpp/all/conanfile.py
@@ -1,6 +1,7 @@
 import os
 from conans import CMake, ConanFile, tools
 from conans.errors import ConanInvalidConfiguration
+from conans.tools import Version


 class PahoMqttCppConan(ConanFile):
@@ -11,7 +12,7 @@ class PahoMqttCppConan(ConanFile):
     license = "EPL-1.0"
     description = "The open-source client implementations of MQTT and MQTT-SN"
     exports_sources = ["CMakeLists.txt", "patches/*"]
-    generators = "cmake", "cmake_find_package"
+    generators = "cmake"
     settings = "os", "arch", "compiler", "build_type"
     options = {"shared": [True, False],
                "fPIC": [True, False],
@@ -46,8 +47,11 @@ class PahoMqttCppConan(ConanFile):


     def requirements(self):
-        self.requires("paho-mqtt-c/1.3.8")
-
+        if Version(self.version) < "1.2.0":
+            self.requires("paho-mqtt-c/1.3.6")
+        else:
+            self.requires("paho-mqtt-c/1.3.8")
+        
     def source(self):
         tools.get(**self.conan_data["sources"][self.version])
         extracted_dir = self.name.replace("-", ".") + "-" + self.version
@@ -66,6 +70,8 @@ class PahoMqttCppConan(ConanFile):
         return self._cmake

     def build(self):
+        if Version(self.version) >= "1.2.0" and Version(self.deps_cpp_info["paho-mqtt-c"].version) < "1.3.8":
+            raise ConanInvalidConfiguration("The project {}/{} requires paho-mqtt-c >= 1.3.8".format(self.name, self.version))
         for patch in self.conan_data["patches"][self.version]:
             tools.patch(**patch)
         cmake = self._configure_cmake()
```

Reply:
First I am sorry for the delay, this is a complicated issue, and the solution is not trivial, just needed to have some free time to look at it.

> Being a conan build newb

Welcome to the club 🤣 💟 

***

> The -c package_info is being called by the -cpp build and the cmake Find file is being placed in the -cpp build directory.

Yes.



> The -cpp build then uses those generated cmake find files to locate the eclipse-paho-mqtt-c library files.

I am not 100% sure that is the case 🙊 I think it's using the file included in the -cpp project **not** the conan ones

You're comments seem to confirm I suspicion!

> Is it the generators in the -cpp recipe doing this? - generators = "cmake", "cmake_find_package"

Yes.

> I can get it to build with generators = "cmake", and this doesn't generate the Findeclipse-paho-mqtt-c.cmake file. Is there more going on?

I am afraid so...

Reply:
Changed by https://github.com/eclipse/paho.mqtt.c/commit/f875768984574fede6065c8ede0a7eac890a6e09
and https://github.com/eclipse/paho.mqtt.c/commit/c116b725fff631180414a6e99701977715a4a690

I think I found a solution....